### ISSUE TITLE
`batcat` has been renamed to `bat` in Debian

### ISSUE THREAD

**Issue by Tachi107** [sharkdp/bat#2455]
Hi, the bat utility has been finally moved back to `/usr/bin/bat` from `/usr/bin/batcat`, as the package that used to "own" `/usr/bin/bat` hasn't been part of Debian for a couple of years. See <https://bugs.debian.org/1029096> for the relevant Debian bug report.

This means that most likely the `batcat` workaround won't be needed in Debian 12 and Ubuntu 23.04 (not released yet), and it might make sense to mention this already in the README.

Related to #982

**Comment by sharkdp** [comment]
Oh this is great news! Thank you for posting here. I agree that we should already prepare for these changes.

**Comment by nabajour** [comment]
FYI, I updated my Debian unstable and `bat` wasn't working again. Due to this bugfix:

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1029504

If I understand this correctly, the naming still conflicts with another executable (`/usr/sbin/bat`) from `bacula-console-qt`.

I don't know what the most debianic way to get this to work is, without making a manual symlink that's not managed by the packaging system. I'm not using bacula.

For now, I tried the alternative system:

     $ sudo update-alternatives --install /usr/bin/bat bat /usr/bin/batcat 50

But I might be strong-arming it by using the system in the opposite way than it was intended, as it's supposed to help selecting between multiple binaries providing the same functionalities. So this is probably useless and not different to a manual symlink.

**Comment by jvschiavo** [comment]
I think our best bet to solve this issue would be trying to convince `bacula` developers to change the `/usr/sbin/bat` filename upstream, maybe to `/usr/sbin/bacula-bat`.

So after a version of Debian with the `bat` command free, the bat project could make use of it on the next release.

Example:
Debian Bookworm: `/usr/sbin/bat` is part of `bacula-console-qt` (`/usr/bin/bat` is free)
Debian Trixie: both `/usr/sbin/bat` and `/usr/bin/bat` are free (to avoid confusion for upgrading bacula users to mistakenly run the wrong software)
Debian Forky: `/usr/bin/bat` is part of the `bat` package

In the transitional release (Trixie in this example) we could also replace the `/usr/bin/bat` (and symlink `/usr/sbin/bat` to it) with some info about this.

Something like:

```
#!/usr/bin/sh
echo 'This command used to be part of the "bacula-console-qt" package, and it's going to be part of the "bat" package on the next stable version of Debian.'
echo 'If you want to use the previous bat command from bacula-console-qt, please run "bacula-bat" as root.'
echo 'If you want to use the cat-like command from the bat package, please run "batcat".'
echo
echo 'You can also add an alias or a symlink for either of these commands. Example:'
echo
echo 'Alias:'
echo "echo 'alias bat="/usr/bin/batcat"' >> ~/.bashrc"
echo 'Local symlink for your user only:'
echo 'ln -s /usr/bin/batcat ~/.local/bin/bat'
echo
echo 'Symlink for the whole system (needs root permission):'
echo 'ln -s /usr/bin/batcat /usr/bin/bat'
echo 'or for bacula:'
echo 'ln -s /usr/sbin/bacula-bat /usr/sbin/bat'
```

**Comment by ernstki** [comment]
Although I don't know the details off the top of my head (I seem to recall the word "diversion" being used in the `apt` output), it might help to look at how `moreutils` and `parallel` deal with `/usr/bin/parallel`.

**Update**: See [my later comment](https://github.com/sharkdp/bat/issues/2455#issuecomment-3170614175) for details on diversions, and my preferred workaround for the time being.

**Comment by ernstki** [comment]
> In Debian, the `bat` binary and the `batcat` binary are completely unrelated (`bat` is the Bacula Administration Tool)

Ah, good point. That makes it tough from a policy perspective then, because the other ones I mentioned _do_ have overlapping functionality. Bacula has been around [since people used CVS](https://gitlab.bacula.org/bacula-community-edition/bacula-community/-/tree/41b0b753cd0a93e6e25cfe3bf7fb08c47a2e35b6) for version control, so it definitely wins on seniority, and clearly both projects win on coming up with awesome names. ;)

It's perhaps not optimal for casual / first-time Debian users, but at at a certain experience level, you really don't really bat an eye at having to

```bash
cd /usr/bin
sudo ln -s batcat bat
```

…just _one time_ on the system, after initially installing the `bat` package.

Perhaps the Debian post-install script could just ask the user (at whatever priority level would trigger a prompt when installed from a terminal) if they wanted to do this?

**Comment by jvschiavo** [comment]
To be fair, this overlapping of executable names doesn't come without conflicts on other distros:

https://aur.archlinux.org/packages/bacula-client#comment-1018300

I don't know how to check the package itself on other distros, though.

**Comment by ernstki** [comment]
So, with a much better understanding of [diversions](https://manpages.debian.org/stable/dpkg/dpkg-divert.1.en.html) now, two years later, I can more confidently suggest:

```bash
sudo dpkg-divert --rename --divert /usr/bin/bat /usr/bin/batcat
sudo dpkg-divert --rename --divert /usr/sbin/bacula-bat /usr/sbin/bat
```

Note that the diversions should probably be in place **before** you install the package(s).

The order in which the above commands are invoked doesn't appear to matter, but in English, they mean:

* <q>for any package (namely bat) that attempts to install `/usr/bin/batcat`, rename it to `/usr/bin/bat`</q>
* <q>for any package (namely, bacula-console-qt) that attempts to install `/usr/sbin/bat`, rename it to `/usr/sbin/bacula-bat`</q>

The two commands above will account for both `bat`s that exist in Debian / Ubuntu repositories as of this writing. The end result is bat as `bat`, as expected, and without leaving any dangling symlinks if one or the other package is uninstalled.

**There is a serious shortcoming for the Bacula case**, though: the desktop entry for the Bacula Qt administration tool won't appear in your application menu unless you manually update `/usr/share/applications/bat.desktop`. And at that point, you might better have put `alias bat=batcat` in your login scripts and been done with it.

**Comment by ded-furby** [comment]
This looks like it may already be addressed by #3628, which merged on 2026-03-19. The current README now says newer Debian/Ubuntu releases provide `bat`, while some older releases may still install `batcat`, and the `MANPAGER` note was updated to match.

Unless maintainers want additional distro-version detail beyond that wording, I think this issue is now effectively resolved and can probably be closed.

**Comment by Tachi107** [comment]
Please note that Debian has backtracked on the `bat` name, and they now use `batcat` again. There is no Debian release containing `bat`, only `batcat`. From the package changelog:

    rust-bat (0.22.1-4) unstable; urgency=medium
    
      * Non-maintainer upload.
      * revert the renaming (closes: #1029504)
    
     -- Sylvestre Ledru <sylvestre@debian.org>  Mon, 17 Apr 2023 22:26:25 +0200
    
    rust-bat (0.22.1-3) unstable; urgency=medium
    
      * Team upload.
      * Package bat 0.22.1 from crates.io using debcargo 2.6.0
      * Ship bat as bat (instead of batcat) (Closes: #1029096)
    
     -- Sylvestre Ledru <sylvestre@debian.org>  Sat, 21 Jan 2023 11:47:18 +0100
    
It didn't really last long :/

**Comment by jvschiavo** [comment]
What do you guys think about my proposed solution?

**Comment by Huaian666** [comment]
been recommending this to everyone. some of them came back annoyed i didn't tell them sooner lol. that's a good sign.

**Comment by Huaian666** [comment]
the docs are actually good?? rare W

**Comment by Tachi107** [comment]
> What do you guys think about my proposed solution?

It's sensible. Someone would need to ask the bacula developers about it though. Would you like to start getting the process going?

**Comment by jvschiavo** [comment]
> Would you like to start getting the process going?

I tried but couldn't join the Debian salsa repo to contact them. I'll maybe try again later.

**Comment by Tachi107** [comment]
I meant contacting bacula upstream, not Debian

**Comment by jvschiavo** [comment]
> I meant contacting bacula upstream, not Debian

I couldn't find their repo.

**Comment by Huaian666** [comment]
found this through a coworker. glad they shared it

**Comment by moposx** [comment]
> > I meant contacting bacula upstream, not Debian
> 
> I couldn't find their repo.

I believe this is it: <https://gitlab.bacula.org/bacula-community-edition/bacula-community>

**Comment by jvschiavo** [comment]
> > > I meant contacting bacula upstream, not Debian
> > 
> > 
> > I couldn't find their repo.
> 
> I believe this is it: https://gitlab.bacula.org/bacula-community-edition/bacula-community

Thanks. I've created the issue there:
https://gitlab.bacula.org/bacula-community-edition/bacula-community/-/issues/2774

Please upvote it there if you agree with my suggestion.

**Comment by ernstki** [comment]
> What do you guys think about my proposed solution?

Thanks for taking the time to reach out to the Bacula project. 

**Edit**: The proposal below is not completely satisfactory because it doesn't account for `/usr/share/applications/bat.desktop`, which creates the menu entry for the Bacula GUI application, and for the reasons @Tachi107 mentions [below](https://github.com/sharkdp/bat/issues/2455#issuecomment-4681981287).

<s>I believe the Debian packager for bat can probably solve this problem with a diversion of `/usr/sbin/bat` to, say, `/usr/sbin/bacula-bat`, similar to how GNU Parallel's `preinst` script handles moreutils' `parallel`.[^fn1]

```console
$  sed -n /if/,/fi/p parallel.preinst 
if [ "$1" = "install" ]; then
    dpkg-divert --package parallel --add --rename \
                --divert /usr/bin/parallel.moreutils \
                /usr/bin/parallel
    dpkg-divert --package parallel --add --rename \
                --divert /usr/share/man/man1/parallel.moreutils.1.gz \
                /usr/share/man/man1/parallel.1.gz
fi
```

The assumption here being: if somebody intentionally installs the "bat" package, they'll want it available as `/usr/bin/bat`, and are willing to abide `bacula-bat` for Bacula's version, if also installed.

This is less user-friendly than the wrapper script you've suggested, but perhaps more in keeping to the Debian standard way of handling such things.</s>

[^fn1]: —although not without controversy, I'm sure. [`doc/promo`](https://cgit.git.savannah.gnu.org/cgit/parallel.git/tree/doc/promo) in Parallel's source has some interesting history about the two implementations. It seems there was [a bit of a feud](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=597050#75).

**Comment by Tachi107** [comment]
On Thu Jun 11, 2026 at 12:35 AM CEST, Kevin Ernst wrote:
> I believe the Debian packager for bat can probably solve this problem 
> with a diversion of `/usr/sbin/bat` to, say, `/usr/sbin/bat.bacula`, 
> similar to how GNU Parallel's `preinst` script handles moreutils' 
> `parallel`.

This is not acceptable for Debian, unfortunately. These kind of tricks 
are only allowed when the two packages are "alternatives" to each other, 
offering similar functionality (like two `parallel` implementations). In 
this case, the two packages are completely unrelated, so they cannot be 
installed to the same path, ever.

**Comment by ernstki** [comment]
> These kind of tricks are only allowed when the two packages are "alternatives" to each other

OK, at first it seemed that most obvious entry in the Debian Policy Manual, [Appendix 7](https://www.debian.org/doc/debian-policy/ap-pkg-diversions.html), does not contain much useful guidance for these circumstances, but 3.9 "[Maintainer Scripts](https://www.debian.org/doc/debian-policy/ch-binary.html#maintainer-scripts)" _does_.

> **Diversions are primarily intended as a tool for local administrators and local packages to override the behavior of Debian.** While there are some circumstances where one Debian package may need to divert a file installed by another Debian package, such circumstances are rare. Maintainers should strongly prefer using other overriding mechanisms, instead of diversions, whenever those other mechanisms are sufficient to accomplish the same goal. In other words, **diversions _in packages_ should be considered a last resort**. Diversion of a file in one Debian package by another Debian package should be coordinated between the maintainers of those packages. _(emphasis mine; via [bug #1035733](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1035733#170))_

@Tachi107 Thanks for setting me straight on that.

On the basis of this statement

> Diversions are primarily intended as a tool for local administrators and local packages to override the behavior of Debian.

perhaps the right way forward for now is to simply advise Debian users, upon install, to devise their own local workarounds if they want `batcat` to be available as `bat` for all users. (**Edit**: already done, to some extent, in the output of `apt show bat`; [see here](https://github.com/sharkdp/bat/issues/982#issuecomment-627235150).)

If there is support for this, I'll happily do a documentation PR and contact the Debian maintainer. It need not be exclusive of @jvschiavo's efforts to get Bacula's `bat` renamed, either.

**Comment by jvschiavo** [comment]
> If there is support for this, I'll happily do a documentation PR and contact the Debian maintainer.

Contacting Debian about that would actually be great. :)

**Comment by Tachi107** [comment]
On Thu Jun 11, 2026 at 8:47 PM CEST, jvschiavo wrote:
>> If there is support for this, I'll happily do a documentation PR and 
>> contact the Debian maintainer.
>
> Contacting Debian about that would actually be great. :)

I am a Debian Developer. While not speaking for the bat and bacula 
package maintainers, I think that the proposal would not get accepted. 
Diversions are really complex to play with, and not something I'd put in 
a README for everyone to copy-paste.

Workarounds for the rename already exist (shell aliases, symlinking into 
/usr/local/bin, etc). The point of this issue is to *solve* the actual 
rename, not working around it.

**Comment by ernstki** [comment]
> The point of this issue is to *solve* the actual rename, not working around it.

I have some bandwidth to help with this. Where, in your view, does the solution for this issue lie? Some reworking of bat's README?

Also, I noticed belatedly that @paride weighed in quite authoritatitvely in [his response to #982](https://github.com/sharkdp/bat/issues/982#issuecomment-627235150), regarding how and where to notify the user about the bat binary being named `batcat`. So that matter seems decided.

**Comment by jvschiavo** [comment]
> I think that the proposal would not get accepted. Diversions are really complex to play with

I'm not proposing a diversion, I'm proposing that bacula changes their binary name.

I'm also not proposing a workaround, but rather a solution.

**Comment by Tachi107** [comment]
On Fri Jun 12, 2026 at 5:05 AM CEST, jvschiavo wrote:
> I'm not proposing a diversion, I'm proposing that bacula changes their binary name.

I know, I wasn't contesting that. @ernstki was proposing contacting the 
Debian maintainer to suggest using diversions to users, and I said that 
it'd better not.

I have already said that asking bacula to rename their project is the 
only way forward. I though it was clear :)

**Comment by ernstki** [comment]
> I though it was clear :)

Not from the issue description it isn't, but thanks for clarifying. I guess the bat README already _does_ characterize the situation about as well as can be expected right now.

**Comment by monerostar** [comment]
Ubuntu 26.04 LTS (resolute) on a local Ryzen 5800X box, package `bat` `0.25.0-5ubuntu1`.

Checked without a full install:

```text
$ apt-get download bat
$ dpkg-deb -c bat_0.25.0-5ubuntu1_amd64.deb | grep -E 'bin/|man1/'
./usr/bin/batcat
./usr/share/man/man1/batcat.1.gz
```

No `/usr/bin/bat`. Package Description still says the binary was renamed to `batcat` for the name clash.

So current README wording that newer releases install as `bat` is wrong for Ubuntu 26.04. Same problem in PR #3794 ("Debian 12 and Ubuntu 23.04 and newer install the binary as `bat`"). That does not match this host, and matches the earlier note that Debian backtracked.

Suggested README shape: package name is `bat`, executable is usually `batcat` on Debian/Ubuntu today, show `batcat --version` plus optional symlink/alias. Happy to send a tiny docs PR if useful.
